### PR TITLE
Fixed #30613 -- Moved index name validation to system checks.

### PR DESCRIPTION
**Pull request by cansarigol**
Ticket [#30613](https://code.djangoproject.com/ticket/30613#ticket)

### CODE REVIEW

File: django/db/models/indexes.py [hunk 1]
```diff
@@ -118,11 +100,6 @@ def set_name_with_model(self, model):
             column_names[0][:7],
             '%s_%s' % (names_digest(*hash_data, length=6), self.suffix),
         )
-        assert len(self.name) <= self.max_name_length, (
-            'Index too long for multiple database support. Is self.suffix '
-            'longer than 3 characters?'
-        )
```
Comment (felixxm) [inline django/db/models/indexes.py]: This assertion should stay in the `Index` class.

File: django/db/models/indexes.py [hunk 2]
```diff
@@ -33,28 +33,10 @@ def __init__(self, *, fields=(), name=None, db_tablespace=None, opclasses=(), co
...
-    def check_name(self):
-        errors = []
-        # Name can't start with an underscore on Oracle; prepend D if needed.
-        if self.name[0] == '_':
-            errors.append('Index names cannot start with an underscore (_).')
-            self.name = 'D%s' % self.name[1:]
-        # Name can't start with a number on Oracle; prepend D if needed.
-        elif self.name[0].isdigit():
-            errors.append('Index names cannot start with a number (0-9).')
-            self.name = 'D%s' % self.name[1:]
```
Comment (felixxm) [inline django/db/models/indexes.py]: Name corrections should be moved into `set_name_with_model()`, e.g.
```python
if self.name[0] == '_' or self.name[0].isdigit():
    self.name = 'D%s' % self.name[1:]
```

File: docs/ref/checks.txt [hunk 3]
```diff
@@ -313,6 +313,7 @@ Models
   ``<model>``.
 * **models.E032**: constraint name ``<constraint>`` is not unique amongst
   models: ``<model list>``.
+* **models.E033**: index name is not valid ``<index>``.
```
Comment (felixxm) [inline docs/ref/checks.txt]: This doesn't match to checks added in `django/core/checks/model_checks.py`, we should add two checks:

- ``` models.E033: Index names cannot start with an underscore or a number: <index>.```,
- ``` models.E034: Index names cannot be longer than <max_length> characters: <index>.```

File: tests/check_framework/test_model_checks.py [hunk 4]
```diff
@@ -131,6 +131,50 @@ class Model2(AbstractModel):
...
+                    models.Index(fields=['id'], name='_name_starting_with_underscore'),
+                    models.Index(fields=['id'], name='5name_starting_with_number')
+                ]
+        self.assertEqual(checks.run_checks(app_configs=self.apps.get_app_configs()), [
+            Error(
+                "Index names cannot start with an underscore (_): _name_starting_with_underscore.",
+                id='models.E033',
+            ),
+            Error(
+                "Index names cannot start with a number (0-9): 5name_starting_with_number.",
```
Comment (felixxm) [inline tests/check_framework/test_model_checks.py]: Please use single quotes.

**Comment by cansarigol** [comment]
I've applied your comments @felixxm

**Comment by felixxm** [comment]
I moved tests to `invalid_models_tests`, moved checks to `Model.check()`, and pushed some minor edits.

**Pull request merged**